identity-iota/src/credential/errors.rs

```diff
@@ -0,0 +1,127 @@
+  #[non_exhaustive]
+  IncompatibleIssuerDocuments,
+
+  /// Indicates that the credential's issuer could not be parsed as a valid DID.
+  #[error("the credential's issuer property could not be parsed to a valid DID")]
+  #[non_exhaustive]
+  IssuerUrl {
+    source: Box<dyn std::error::Error + Send + Sync + 'static>, /* Todo: Would it be better to use a specific type
+                                                                 * here? */
+  },
+
+  /// Indicates that the presentation's holder could not be parsed as a valid DID.
+  #[error("the presentation's holder property could not be parsed to a valid DID")]
+  #[non_exhaustive]
+  HolderUrl {
+    source: Box<dyn std::error::Error + Send + Sync + 'static>, /* Todo: Would it be better to use a specific type
+                                                                 * here? */
+  },
+
+  /// Indicates an attempt to verify a presentation's signature with a resolved DID document not corresponding to the
+  /// URL of the presentation's holder property.
+  #[error("the provided holder document does not correspond to the presentation's holder property")]
+  #[non_exhaustive]
+  IncompatibleHolderDocument,
+
+  /// Indicates that the presentation's signature could not be verified using the holder's DID Document.
+  #[error("could not verify the holder's signature")]
+  #[non_exhaustive]
+  HolderProof {
+    source: Box<dyn std::error::Error + Send + Sync + 'static>, /* Todo: Would it be better to use a specific type
+                                                                 * here? */
+  },
+  /// Indicates that the structure of the [Credential](identity_credential::credential::Credential) is not semantically
+  /// correct.
+  #[error("the credential's structure is not semantically correct")]
+  CredentialStructure(#[source] identity_credential::Error),
+  /// Indicates that the structure of the [Presentation](identity_credential::presentation::Presentation) is not
+  /// semantically correct.
+  #[error("the presentation's structure is not semantically correct")]
+  PresentationStructure(#[source] identity_credential::Error),
+  /// Indicates that the relationship between the presentation holder and one of the credential subjects is not valid.
+  #[error("expected holder = subject of the credential at position {credential_position}")]
+  #[non_exhaustive]
+  InvalidHolderSubjectRelationship { credential_position: usize },
+  /// Indicates that the presentation does not have a holder.
+  #[error("the presentation has an empty holder property")]
+  MissingPresentationHolder,
+}
+
+// Todo: Consider implementing Error for OneOrMany<E: std::error::Error> to avoid wrapping it in
+// AccumulatedCredentialValidationError
+#[derive(Debug)]
+/// An error caused by a failure to validate a Credential.  
+pub struct AccumulatedCredentialValidationError {
+  pub validation_errors: OneOrMany<ValidationError>,
+}
```

Comment:
Is `OneOrMany` necessary here? I think `Vec<ValidationError>` should be sufficient since there is no enforcement that it is non-empty and special handling for one error vs many doesn't seem required (like it is for (de)serialization in data structures).

Reply:
I don't have a strong opinion on this really. The initial idea was that if one supplies `fail_fast = true` to `validate(credential/presentation, options, fail_fast)` you don't need to pay for allocation upon failure with the `OneOrMany::One` variant.

Reply:
I'm fairly certain that an empty `Vec` is not heap-allocated and since an enum is sized for its largest variant we're not saving any space at all with `OneOrMany::One`, in my understanding.

https://doc.rust-lang.org/beta/alloc/vec/struct.Vec.html#method.new

Reply:
I know that `CompoundPresentationError` uses `Vec` in place of `OneOrMany` however and the two should ideally be consistent. So `OneOrMany` both places or `Vec` both places?

Reply:
> I guess the first allocation for the Vec will be slightly larger than one element 

The first allocation also happens on the heap while with `OneOrMany` there will be no heap allocation at all in the `fail_fast = true` scenario. If you think the simplicity of `Vec` outweighs the performance gains in the unhappy path with fail_fast using `OneOrMany` I will be happy to replace `OneOrMany` with `Vec`.

Reply:
It would avoid other changes like adding `OneOrMany::empty` too, which is not an intended use-case for the struct. I don't believe avoiding a single allocation to be significant here since some of the errors already use `Box` internally.